node/core/provisioner/src/lib.rs

```diff
@@ -0,0 +1,362 @@
+						&& merged_bitfields_are_gte_two_thirds(
+							*core_idx,
+							&signed_bitfields,
+							&occupied.availability,
+						);
+				}
+				if let Some(ref scheduled) = occupied.next_up_on_time_out {
+					return scheduled.para_id == para_id
+						// TODO: this function currently doesn't know about what block we're building
+						//   get it from the runtime? where?
+						&& unimplemented!("occupied.time_out_at == the block we are building")
+						// REVIEW: the guide says we have to ensure that the merged bitfields are < 2/3 in this case,
+						// but intuition suggests that we'd still want it to be >=. Just want a double-check that
+						// there's no typo in the guide.
+						&& !merged_bitfields_are_gte_two_thirds(
+							*core_idx,
+							&signed_bitfields,
+							&occupied.availability,
+						);
+				}
+				false
+			}
+		}
+	});
+
+	// ensure the postcondition holds true: sorted by core index
+	// note: unstable sort is still deterministic becuase we know (by means of `parachains_represented`) that
+	// no two backed candidates remain, both of which are assigned to the same core.
+	backed_candidates.sort_unstable_by_key(|candidate| {
+		let para_id = candidate.candidate.descriptor.para_id;
+		let (core_idx, _) = cores_by_id.get(&para_id).expect("paras not assigned to a core have already been eliminated from the backed_candidates list; qed");
+		core_idx
+	});
+
+	// type ProvisionerInherentData = (Vec<SignedAvailabilityBitfield>, Vec<BackedCandidate>);
+	return_sender
+		.send((signed_bitfields, backed_candidates))
+		.map_err(|_| Error::OneshotSend)?;
+	Ok(())
+}
+
+// The instructions state:
+//
+// > we can only include the candidate if the bitfields we are including _and_ the availability vec of the OccupiedCore
+//
+// The natural implementation takes advantage of the fact that the availability bitfield for a given core is the transpose
+// of a set of signed availability bitfields. It goes like this:
+//
+//   - organize the incoming bitfields by validator index
+//   - construct a transverse slice along `core_idx`
+//   - bitwise-or it with the availability slice
+//   - count the 1 bits, compare to the total length
+//
+// REVIEW: is that interpretation the correct one?
+fn merged_bitfields_are_gte_two_thirds(
+	core_idx: usize,
+	bitfields: &[SignedAvailabilityBitfield],
+	availability: &BitVec<bitvec::order::Lsb0, u8>,
+) -> bool {
+	let mut transverse = availability.clone();
+	let transverse_len = transverse.len();
+
+	for bitfield in bitfields {
+		let validator_idx = bitfield.validator_index() as usize;
+		match transverse.get_mut(validator_idx) {
+			None => {
+				// REVIEW: might it be worth having this function return a `Result<bool, Error>` so that we can more clearly express this error condition?
+				log::warn!(target: "provisioner", "attempted to set a transverse bit at idx {} which is greater than bitfield size {}", validator_idx, transverse_len);
+				return false;
+			}
```

Comment:
I'm fairly confident that this will only ever happen in the event of a programming error, but not confident enough to just do a `.expect("can't happen; qed")` thing. Returning `Result<bool, Error>` also feels odd, but might be the right way to approach this situation.

Reply:
That's the right intuition. We generally don't want the node to go down if the runtime API returns garbage data - just act very bewildered.